astroid/inference.py

```diff
@@ -58,9 +71,24 @@
 from astroid.interpreter import dunder_lookup
 from astroid.manager import AstroidManager
 
+if TYPE_CHECKING:
+    from astroid.objects import Property
+
 # Prevents circular imports
 objects = util.lazy_import("objects")
 
+if sys.version_info >= (3, 8):
+    from typing import TypedDict
+else:
+    from typing_extensions import TypedDict
+
+FunctionDefT = TypeVar("FunctionDefT", bound=nodes.FunctionDef)
+
+
+class NodeContextPair(TypedDict):
```

Comment:
I wasn't so far off with my first guess. Take a look at the `raise_if_nothing_inferred` decorator.
https://github.com/PyCQA/astroid/blob/40e1a64ad67d4613b0bef461b0227653c8f085a1/astroid/decorators.py#L149-L155
It assigns the `StopIteration` error args to the `InferenceError` exception class.
(A `StopIteration` exception is raised with a return inside a generator function.)

We shouldn't use `InferenceError` for the dict itself. Maybe `InferenceErrorTD`? Don't really know, maybe there is an even better name.

Reply:
Hm, but what about the `_cache` then? That does not raise an `InferenceError` right? 
I tested with the `test_extension_modules` test and a breakpoint just before the second return in `infer_functiondef`.

Reply:
Ah, did confuse that with the `_cached_generator` with the `cached` decorator. Just too many decorators 😄

That a look at what is actually stored in the `_cache` dict. Only `[<Property.name ...>]` not the return value. `list(copy)` does iterate over the Generator but implicitly handles `StopIteration` exceptions (with the `return` statements basically are). https://stackoverflow.com/a/16780113

Reply:
> Hm, so what do we want to name this? `InferenceErrorTD`?

As we both have no other idea, I guess it's good enough 🤷🏻‍♂️

Reply:
I mean, `NodeContextPair` makes this somewhat re-usable if we ever need it in another context, but both are fine.

I'll change to `InferenceErrorTD`

Reply:
What about `InferenceErrorInfoTD`? From what I can tell, it's only ever used for the `raise_if_nothing_inferred` decorator and the fallback case even mentions "without any error information".
https://github.com/PyCQA/astroid/blob/2a064bac3430daead3a0c564fb5abf82257654f1/astroid/decorators.py#L147-L158

I would also suggest, to move it to `decorators.py` directly. We might later need to use if in `node_classes.py` for `unpack_infer` and since it's always used in combination with `raise_if_nothing_inferred` the `decorators` file is already imported. So we don't risk any new circular dependencies.

_As we'll be importing it in other files, it should also be public (without the leading `_`)._

Reply:
> _As we'll be importing it in other files, it should also be public (without the leading `_`)._

Doesn't the leading underscore mean it is private for "external" users? As in, to be used internally within this package?
If not, how do we signal that something is used internally but should not be considered public?

Reply:
> If not, how do we signal that something is used internally but should not be considered public?

There's definitely some clarifying to do about that. It's a little fuzzy. I think right now we consider everything "private" unless it's documented (how to run pylint programmatically ?) or widely used by downstream libraries. But we had a lot of regressions downstream with this this attitude (we don't know what is used by downstream libraries, they are a lot of them).

Reply:
> Doesn't the leading underscore mean it is private for "external" users? As in, to be used internally within this package?

Yes, but "external" users in this case are also other modules. Usually, it's better to avoid importing anything with a leading underscore (even from the same package).

> If not, how do we signal that something is used internally but should not be considered public?

Python doesn't really have a good concept for that, at least non that I know of. In general everything should be considered public. Some pypi packages try to split public and private functions and create entirely separate packages. `pytest` is an example here with `_pytest` for internal code. However that is a lot of work, which is not worth it IMO.

> There's definitely some clarifying to do about that. It's a little fuzzy. I think right now we consider everything "private" unless it's documented (how to run pylint programmatically ?) or widely used by downstream libraries. But we had a lot of regressions downstream with this this attitude (we don't know what is used by downstream libraries, they are a lot of them).

Yeah there is a risk of accidental regressions but as far as I remember, those where mostly caused by large scale refactorings rather than actual code changes.

In the end, it's a tradeoff. For methods and attributes I usually add an underscore (when it should be private). Just because it's easy to do and the code is already scoped. For functions only if they are true helper functions used exclusively in that module. Similar for classes. A good analogy here are the TypeVar and TypeAliases. There it does make sense to add one.

Those "soft" rules seem to have worked quite good so far.

Reply:
It's definitely a trade off. Should we document the current way of doing thing it in the dev guide ? Saying that everything is public unless prefixed by an underscore will increase the burden of handling retro-compatibility by a lot.

Reply:
> Yes, but "external" users in this case are also other modules. Usually, it's better to avoid importing anything with a leading underscore (even from the same package).

> In the end, it's a tradeoff. For methods and attributes I usually add an underscore (when it should be private). Just because it's easy to do and the code is already scoped. For functions only if they are true helper functions used exclusively in that module. Similar for classes. A good analogy here are the TypeVar and TypeAliases. There it does make sense to add one.

The best "authoritative" piece in PEP8 on this topic I could find is [this](https://www.python.org/dev/peps/pep-0008/#id46).

For me the benefit of underscoring `astroid` "internals", comes from avoiding downstream regressions. I understand the rule is that you also shouldn't import underscored interfaces from modules within the same package, but when you do and a regression occurs you'll at least be able to fix that yourself.
For me that's a better trade-off than having to tread carefully everywhere because downstream packages might depend on something that is not underscored.

Reply:
> It's definitely a trade off. Should we document the current way of doing thing it in the dev guide ? Saying that everything is public unless prefixed by an underscore will increase the burden of handling retro-compatibility by a lot.

Not adding an underscore doesn't necessarily mean that it's intended to be public. IMO it's better to leave the situation as is and decide from case to case if necessary. That did work so far with only a few exceptions.

The only things I consider to be **truly** public are those in `astroid.__init__` and `astroid.nodes.__init__` (in `__all__`).

To close this, I'm not sure the discussion here is worth it. It's a TypedDict used only for astroid after all. Even without an underscore, external projects are unlikely to use it. They simply don't need to.

Reply:
> The only things I consider to be truly public are those in astroid.__init__ and astroid.nodes.__init__ (in __all__).

Agreed, also numversion in __pkginfo__.